### DISCUSSION TITLE
option for pre_text_length size & post_text_length, line numbers and highlight formats in context

### DISCUSSION THREAD

**Discussion by kocser2** [ccbogel/QualCoder#732] [Ideas]
Hello, 
first of all, this is a really great tool. Really! I will write an article about this tool after I finish my work with this program.

Is it possible to add an option to customize both variables: pre_text_length, post_text_length ?
I think this would help a lot to reduce the amount of generated context.

Additionally, an option for the highlighted text within the context. Right now it's just bold. But a few more options would be really great here.

As a third suggestion, to round everything up, would be an option to add line numbers to the text data and apply them also to the report. This way, with large data, referencing to the coded parts is easier.

I hope these are options that find approval.

Many thanks and greetings
Sergej

![Screenshot 2022-12-03 164141](https://user-images.githubusercontent.com/27814189/205450349-89a6f1d2-1957-44be-bed7-273ca3bd80b0.jpg)
![Screenshot 2022-12-03 165742](https://user-images.githubusercontent.com/27814189/205450351-cb42989d-a2a5-420a-8fd4-f44ce3b33e2a.jpg)

**Comment by ccbogel** [comment]
Some good suggestions. I do not think i can do line numbers,  but the other changes I have now updated for 3.2 code.

**Reply by kocser2** [reply]
Thank you for the quick reply. Very nice. 
I think proper pre-formatting of the data on the user side will be necessary for later references. For line numbers, you would probably need to split the text into lines with X number of characters/words. Then insert numbers in front and export the text back to a text file. With certain code words indicating the start of a new dialog or something. Therefore I found it very useful to work with html, because there everything is coded separately and that helps to find certain string parts. 
Also, the html file already has the line splitting done. The only thing that was still a problem with me as a German is when I wanted to save the text from QualCoder to html without context, then it didn't recognize my special character encoding (umlauts). However, when exporting with context, it's not a problem, although the html code doesn't seem to differ. Weird.

**Comment by ccbogel** [comment]
"probably need to split the text into lines with X number of characters/words" - I will not do this - this is a very error prone approach. I could apply a reference number to each heading 1 [view] ... 2 [view] , usw.

The umlauts problem - I think this is a problem with the Qt framework - its works when other text is present (I even tested with Japanese), but not alone:

![pic1](https://user-images.githubusercontent.com/5834638/205763506-6ab893f5-d86a-4622-b37a-146e63218000.png)
It does not work when it is the text alone:
![pic2](https://user-images.githubusercontent.com/5834638/205763674-1544afce-52ac-42dc-80b3-67fc319a00be.png)
You could do a find and replace on the html file

**Reply by kocser2** [reply]
Interesting. Do you think this is a bug of Qt? Or is that some kind of AI-assisted generation that needs some context to work properly?
And how does the HTML export work? Is that a Qt feature? When I try exporting from other programs like Word or Adobe Acrobat, the HTML code looks very messy and not suitable for simple string manipulation. Your HTML code is perfect. All the formatting information regarding a line of text is on the same line in the HTML code. It is clean and clear. Great for string manipulation.

the thing with the line numbers. for me it works if I insert them into the text file with word, then print the file as pdf and export it again to word. then the line numbers are inserted before each line and can be easily manipulated. and the line splitting is also done automatically during this process. QualCoder preserves my pre-formatting perfectly so the result looks like in the picture below.

![Screenshot 2022-12-06 030734](https://user-images.githubusercontent.com/27814189/205791202-4e9ba372-f5e4-4ce2-8477-30e08abd5554.jpg)

**Comment by ccbogel** [comment]
I think this is a bug of Qt. (There is no AI).

Ah, now I understand how you got the line numbers in there. You are the first person to want line numbers. Your document manipulations seem to be the best approach for this. I notice with NVivo, (probably one of the most comprehensive QDA software), this is not a function either and a user of NVivo does similar as you do: https://forums.nvivobyqsr.com/topic/4067-line-numbers/

The HTML code is created by Qt6 in only a couple of lines of code:

tw = QtGui.QTextDocumentWriter()
tw.setFileName(filepath)
tw.setFormat(b'HTML')
tw.write(self.ui.textEdit.document())

**Reply by kocser2** [reply]
Ok, if it is a bug, hopefully, it will be fixed eventually. When I have time, I will look for a small tool that can do my actual workaround. Or try to write one myself.

**Comment by kocser2** [comment]
I just watched the webinar on the presentation of NVivo. This program is huge, but the text coding in it is really cumbersome. For long interviews, you'll definitely need a replacement mouse because the first mouse has broken buttons from all the clicking. With QualCoder this is really easy and very fast.

**Comment by ccbogel** [comment]
Well I am glad QualCoder for $0 is better than Nvivo for $100's for text coding.

**Reply by kocser2** [reply]
$100 for a student license, but I would have had to pay even the full $1200 price. Before I found QualCoder, I wanted to write some scripts and tools to help me code the texts faster. But luckily you've already done all the work. 

So IF I get paid for my QDA work, hopefully before Christmas, I would gladly make a donation to you for the work and write an article somewhere on linked-in, xing, reddit for example to help promote QualCoder and further support its development.

**Reply by ccbogel** [reply]
I dont know how to set up donations yet. But writing articles and promotion would be good.

**Comment by kocser2** [comment]
How difficult would it be, or what would it take, to implement code highlighting like in a software code editor and display it on the side next to the scroll bar? 

![Screenshot 2022-12-07 162643](https://user-images.githubusercontent.com/27814189/206220793-26b971a4-3fdc-4bbc-a187-34efc213263b.jpg)

**Reply by ccbogel** [reply]
I have a feeling it would be difficult as resizing changes the text positioninh. But i have never tried it.

**Reply by kocser2** [reply]
How about just displaying the highlight next to the scroll bar without the miniature Text View when you click on a codeword and press a key to highlight all coded text locations in the text?

**Comment by kocser2** [comment]
I noticed that my hard drive works after each character I type when I write notes into the journal, for example, (not Memos, that are treated differently I guess). And this writing process takes some time, so I guess typing on older systems can seem a bit laggy. How about waiting for a "space" as an indicator to initiate a write? Or just every 10 seconds, or so. Have you encountered this kind of performance issue? Or is it because the version of QualCoder I'm using is a portable (exe version)?

**Reply by ccbogel** [reply]
HI. I have not encountered slowness yet. It does a save to the database after every character is typed (or deleted). So it is automatically saving as you go. I can consider changing options to how this works.

**Reply by ccbogel** [reply]
Are you finding it very laggy?

**Reply by kocser2** [reply]
in this particular situation (typing in the journal), it is "slow" and laggy yes, especially if you are a fast typer :)

**Comment by ccbogel** [comment]
Oh one more thing: are you typing in the journal in this location: Manage Files > Journals

**Reply by kocser2** [reply]
Yes, is this not the intended way?

**Reply by ccbogel** [reply]
yes that is the usual way. - That is the one I will work on to change.
There is a hidden panel on the right hand side of the code text section. if you pull it out, you have access to, amongst other things, the most recent journal (by date).

**Reply by ccbogel** [reply]
OK. I have delayed the database updated to every 1.5 seconds. this should hide any lag. Future issues please post them on the issues page. as the discussion sections gets pretty long.
